src/components/CategoryNavContainer/CategoryNav/CategoryNav.js

```diff
@@ -1,34 +1,61 @@
-import React from 'react';
+import React, { Component } from 'react';
 import PropTypes from 'prop-types';
+import classnames from 'classnames';
+import camelCase from 'lodash/camelCase';
 import { NavLink } from 'react-router-dom';
 import style from './CategoryNav.scss';
 import { NO_CATEGORY } from '../../../constants';
 
-const CategoryNav = props => (
-  <div>
-    <NavLink
-      className={style.link}
-      activeClassName={style.link__active}
-      to={`/${props.cityId}/${NO_CATEGORY}`}
-    >Overview</NavLink>
+class CategoryNav extends Component {
+  componentDidMount() {
+    this.scrollToActiveNavLink();
+  }
 
-    {props.categories.map(category => (
-      <NavLink
-        className={style.link}
-        activeClassName={style.link__active}
-        key={category.id}
-        to={`/${props.cityId}/${category.id}`}
-      > {category.name} </NavLink>
-    ))}
-  </div>
-);
+  componentDidUpdate() {
+    this.scrollToActiveNavLink();
+  }
+
+  scrollToActiveNavLink() {
+    if (this.activeLinkEl) {
+      this.el.scrollLeft = this.activeLinkEl.parentElement.offsetLeft - 20;
+    }
+  }
+
+  render() {
+    return (
+      <div ref={(el) => { this.el = el; }} className={style.categoryNav}>
+        <NavLink
+          className={classnames(style.link, style.link__overview)}
+          activeClassName={style.link__active}
+          to={`/${this.props.cityId}/${NO_CATEGORY}`}
+        >Overview</NavLink>
+
+        {this.props.categories.map(category => (
+          <NavLink
+            className={classnames(style.link,
+              style[`link__${camelCase(`category ${category.id}`)}`])}
+            activeClassName={style.link__active}
+            key={category.id}
+            to={`/${this.props.cityId}/${category.id}`}
+          >
+            <span ref={(el) => {
```

Comment:
I'm not very happy with this. The problem I faced was that NavLink, as a stateless functional component, cannot set a reference via `ref` (it just sets a `activeLinkEl` to null). I need to have a reference because I can't calculate how far to scroll until everything is rendered - without duplicating the same calculations are being done by the DOM, which would be the worst possible solution. (See Nick's 'mobile nav' mockups in InVision for the kind of scrolling that needed to be implemented.)

So, I am nesting the label inside a <span>, simply to get a reference, and then checking the parent element in the `scrollToActiveNavLink` function. This is gross, but I can't see another way to do it. Suggestions welcome.

Reply:
I'm not sure that this is better, but you can get away without using refs by just responding to an `onClick` event passed into `<NavLink>`. To scroll to the current one when the page loads you can do the pretty nasty thing of looking it up by class name. It's bad form, but I think if you set it as a property not much can go wrong.

```jsx

function scrollElementIntoView(el) {
  el.parentElement.scrollLeft = el.offsetLeft - 40; // eslint-disable-line no-param-reassign
}

class CategoryNav extends Component {
  constructor(props) {
    super(props);

    this.activeElementClassName = style.link__active;
  }

  componentDidMount() {
    const activeEl = document.querySelector(`.${this.activeElementClassName}`);
    scrollElementIntoView(activeEl);
  }

  render() {
    return (
      <div className={style.categoryNav}>
        <NavLink
          className={classnames(style.link, style.link__overview)}
          activeClassName={this.activeElementClassName}
          to={`/${this.props.cityId}/${NO_CATEGORY}`}
        >Overview</NavLink>

        {this.props.categories.map(category => (
          <NavLink
            key={category.id}
            className={classnames(style.link,
              style[`link__${camelCase(`category ${category.id}`)}`])}
            activeClassName={this.activeElementClassName}
            to={`/${this.props.cityId}/${category.id}`}
            onClick={(e) => {
              scrollElementIntoView(e.target);
            }}
          >
            {category.name}
          </NavLink>
        ))}
      </div>
    );
  }
}
```

Reply:
Ahhh I hadn't thought of pulling the class name out of `style`. I thought the fact CSS Modules modifies the class names would preclude that. I prefer your implementation; at least there's not that extra HTML element.